agent/qrexec-agent-data.c

```diff
@@ -500,6 +509,209 @@ static int process_child_io(libvchan_t *data_vchan,
+            size_t const total_path_length = directory_length + (index ? service_length : path_length);
+            memcpy(remote.sun_path, paths[i].string, directory_length);
+            memcpy(remote.sun_path + directory_length, realcmd, path_length);
+            remote.sun_path[total_path_length] = '\0';
+
+            if (0 == connect(s, (struct sockaddr *) &remote,
+                        (socklen_t)(offsetof(struct sockaddr_un, sun_path) + total_path_length + 1))) {
+                *stdout_fd = *stdin_fd = s, *stderr_fd = -1, *pid = 0;
+                set_nonblock(s);
+                return 0;
+            }
+            switch (errno) {
+            // These cannot happen
+            case EFAULT:       // all of our parameters are in valid memory
+            case EINVAL:       // we passed valid parameters
+            case ENOTSOCK:     // ditto
+            case ENETUNREACH:  // cannot happen for AF_UNIX
+            case EADDRNOTAVAIL:// cannot happen for AF_UNIX
+            case EINPROGRESS:  // this socket is blocking
+            case EBADF:        // `s` was created by a call to `socket()`
+            case EAFNOSUPPORT: // the kernel supports AF_UNIX
+            case EALREADY:     // ditto
+                abort();
+            // These should not happen
+            case EAGAIN:       // ditto
+#if EAGAIN != EWOULDBLOCK
+            case EWOULDBLOCK:  // ditto
+#endif
+                // Probably a misbehaving FUSE filesystem
+                break;
+            case EINTR:
+                // Interrupted by a signal - retry
+                --i; // to retry loop iteration
+                continue;
+            case ENOENT:
+            case EISDIR:
+            case ENOTDIR:
+            case EIO:
+                // These errors could also happen with `execve()`
+                break;
+            case EPROTOTYPE:
+            case ETIMEDOUT:
+                // Socket server problem.  Fail the whole connection.
+                // (We do not want to fall back because the user may have overriden
+                // behavior for security reasons, and because fail-fast is much
+                // easier to debug).
+                goto fail;
+            case EACCES:
+            case EPERM:
+            case ECONNREFUSED: {
+	        if (access(remote.sun_path, X_OK) != 0) {
+                    fprintf(stderr, "cannot execute %s: no execute permission\n", remote.sun_path);
+                    break;
+                }
+                struct stat stat_buf;
+                if (stat(remote.sun_path, &stat_buf) != 0) {
+                    fprintf(stderr, "cannot execute %s: cannot stat\n", remote.sun_path);
+                    break;
+                }
+                if (!S_ISREG(stat_buf.st_mode)) {
+                    fprintf(stderr, "cannot execute %s: not a regular file\n", remote.sun_path);
+                    break;
+                }
+                if ((stat_buf.st_mode & 022)) {
```

Comment:
I don't think such restriction should be hardcoded. And definitely not without matching documentation change.

Reply:
Having a file that is both executable, and group- or world- writeable, is bad practice on \*nix systems.  That said, I agree that this change is probably not useful.